### DISCUSSION TITLE
Polling from stdin getting "stuck".

### DISCUSSION THREAD

**Discussion by fiatjaf** [libuv/libuv#3732] [Q&A] [answered]
I'm using [poll](http://docs.libuv.org/en/v1.x/poll.html) to read messages written to stdin. Basically calling `uv_poll_init` with fd 0 and then `uv_poll_start` with `UV_READABLE`.

There is a program A (outside my control) that starts my program B and writes to my program's stdin. And everything pretty much works.

However sometimes the reads get "stuck", i.e. a message that should have been written by A does not reach B -- the poll doesn't get triggered. I wait seconds, minutes, and nothing. Then if I force program A to write something else to B's stdin, then B gets triggered and is able to read the first and the second messages.

I am very much confused here, is there something I should be doing that I am not? Is this a bug in program A?

Here is my code, if that helps: https://github.com/fiatjaf/poncho/blob/f8e28fbd32d922bd4cb562c251066169fa2e9513/src/main/scala/StdinReader.scala (it's Scala Native bindings).

**Comment by vtjnash** [comment]
`stackalloc` sounds like a bad idea? you probably need malloc there

**Reply by fiatjaf** [reply]
Thank you. I've changed to `malloc` but it didn't make any difference.

**Comment by fiatjaf** [comment]
I am wondering if this polling thing is the right approach. Is it? I guess it is, but could it also work if I opened stdin as a file and read from that? Or would that be a synchronous operation? (Sorry, I'm a newbie in these low-level matters.)

**Comment by fiatjaf** [comment]
Here is a minimal reproducible example: https://github.com/fiatjaf/libuv-stdin-poll-bug (I should probably have written it in C but I don't know how to do that. I will try.)

**Reply by fiatjaf** [reply]
I have written the subprocess code in C now:

```c
#include <stdio.h>
#include <stdlib.h>
#include <uv.h>

void cb(uv_poll_t *handle, int status, int events) {
  char *line = NULL;
  size_t len = 0;
  getline(&line, &len, stdin);
  fprintf(stdout, "GOT LINE: `%s`\n", line);
  fflush(stdout);
}

int main() {
  fprintf(stdout, "STARTING\n");
  fflush(stdout);
  uv_poll_t* handle = (uv_poll_t*) malloc(uv_handle_size(UV_POLL));
  if (0 != uv_poll_init(uv_default_loop(), handle, 0)) { return 1; };
  if (0 != uv_poll_start(handle, UV_READABLE, cb)) { return 2; };
  uv_run(uv_default_loop(), UV_RUN_DEFAULT);
  return 0;
}
```

The results are the same:

https://user-images.githubusercontent.com/1653275/187974402-5774b35b-55ee-4011-b839-c28c44e4d8f3.mp4

**Reply by bnoordhuis** [reply]
Minor thing but this:
```c
uv_poll_t* handle = (uv_poll_t*) malloc(uv_handle_size(UV_POLL));
```
Can be written more efficiently and robustly as:
```c
uv_poll_t* handle = malloc(sizeof(*handle));
```

**Reply by fiatjaf** [reply]
Thank you. I'll try to remember that.

**Comment by vtjnash** [comment]
did you check if you flushed your writes? this sounds like you have a problem in either your go code or in the scala bindings, not libuv

**Reply by fiatjaf** [reply]
I think Go flushes the writes automatically. If it doesn't then I don't know how to proceed because it doesn't expose that API anyway. If the problem are the flushes would you expect this behavior to only happen if multiple sequential writes are made without some delay between them?

https://user-images.githubusercontent.com/1653275/187973694-0069c64a-a200-4ef7-95d5-25baa51512ce.mp4

**Reply by vtjnash** [reply]
this forum isn't really a free-code-help place. I would suggest running your code in strace and rr to find your bug first, then report back if it still seems to be in libuv

**Reply by fiatjaf** [reply]
Sorry, sir, but I'm not asking you to fix my code, I am trying to ask a simple question: is this stuck behavior expected or not and if not how to proper use the library.

All the libuv pages point to this forum for _the place to ask for help_, and that's what I am doing.

If I knew this was for certain a bug in libuv as you want me to I wouldn't be asking for help, I would open an issue describing the bug or fix it and submit a patch.

**Reply by vtjnash** [reply]
it is not expected. generally I would think it would only be line-buffered by default

**Reply by fiatjaf** [reply]
Thank you very much, now I will proceed to try to learn how to use these things, strace and rr.

**Comment by fiatjaf** [comment, marked as answer]
So what I've discovered using `strace` was that when the "writer" process was faster than the "reader" process the lines were all there waiting to be read, but the "reader" process only read one line _and discarded the others_.

So apparently this is totally unrelated to libuv.

This happened both in the Scala land and with `getline`. It can be dealt with by reading from stdin character by character manually then splitting the result by newlines manually.

The only question I have now (and it is off-topic here, I know) is: is this expected from the use of `getline`?

**Reply by fiatjaf** [reply]
Actually this doesn't make sense. Because on the next poll trigger the supposedly-discarded lines are available to be read again. Why? I am confused now, and yet my fix seems to work.

**Reply by vtjnash** [reply]
It sounds like getline has an internal buffer, and you are failing to fully drain it
